### ISSUE TITLE
MacOS Mojave - Cannot link `ld: library not found for -lcf4ocl2`

### ISSUE THREAD

**Issue by praisethemoon** [fakenmc/cf4ocl#37]
Hello,

I have tried both compilation from scratch and `https://formulae.brew.sh/formula/cf4ocl`, yet. cmake. cannot link the library. Am I missing something?

Cheers.

**Comment by fakenmc** [comment]
Hi,

Are you trying to compile cf4ocl or something that depends on cf4ocl?

1. If you're trying to compile cf4ocl, try disabling the build of examples and tests in CMake.
2. If you're trying to compile something which depends on cf4oc, it's possible that you also have to use the `-L` flag to specify the location of the library. Otherwise try using `pkg-config` to generate the flags, as that's the [preferred way](https://github.com/fakenmc/cf4ocl/wiki/Using-cf4ocl-in-a-new-project) to do so anyway.

I don't have access to MacOS Mojave right now, maybe in about two or three weeks, so this is just guesswork. Let me know how it goes.

**Comment by praisethemoon** [comment]
Thank you for your quick reply!

I have compiled the library successfully and started working on a skeleton project, I have the following setup:
CMakeLists:

```
# Find required libraries
find_package(PkgConfig REQUIRED)
pkg_check_modules(CF4OCL2 REQUIRED cf4ocl2>=2.0.0)

# Library include directories
include_directories(${CF4OCL2_INCLUDE_DIRS})

target_link_libraries(nn ${CF4OCL2_LIBRARIES})
```

I still get same error about `ld: library not found for -lcf4ocl2`.

Any suggestion? I have removed the installed files of the library from the manual compilation (which was successful however) and used the brew formula and I still have the same issue.

if you have anything in mind you want to try, let me know 👍

**Comment by praisethemoon** [comment]
`nano /usr/local/lib/pkgconfig/cf4ocl2.pc` gives:

```
refix=/usr/local/Cellar/cf4ocl/2.1.0
exec_prefix=${prefix}
libdir=${prefix}/lib
includedir=${prefix}/include

Name: cf4ocl2
Description: C Framework for OpenCL
URL: http://fakenmc.github.io/cf4ocl/
Version: 2.1.0
Requires.private: glib-2.0 >= 2.32.1
Libs: -L${libdir} -lcf4ocl2
Cflags: -I${includedir} -I/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/OpenCL.framework
```

And contents of `/usr/local/Cellar/cf4ocl/2.1.0/lib/` is :

```
libcf4ocl2.2.1.0.dylib	libcf4ocl2.2.dylib	libcf4ocl2.dylib	pkgconfig
```

**Comment by fakenmc** [comment]
Yeah, you seem to have everything perfectly as it's supposed to. All the stuff seems to be in the right folders and `pkg-config` is correctly pointing towards those folders.
You could try using [cf4ocl without `pkg-config`](https://github.com/fakenmc/cf4ocl/wiki/Using-cf4ocl-in-a-new-project-without-pkg-config), but honestly, I don't have a clue what's going on at this time.
I will definitely try and check this out when I have access to a macOS machine (in 2-3 weeks time). In the meanwhile, if you happen to find out what's wrong do let me know. Sorry I can't be more helpful at this time.

**Comment by praisethemoon** [comment]
Guess I will have to be patient. Alright, thanks a lot. I will see if I can get to work without `pkg-config`.

**Comment by praisethemoon** [comment]
Hello again, I have solved the problem, by using the following CMake Settings:
CF4CL2 settings
```
SET(CF4OCL2_INCLUDE_SEARCH_PATHS
		/usr/include/isaac
		/usr/include
		/usr/local/include
		/opt/isaac/include
		/usr/local/include
		/usr/local/Cellar/cf4ocl/2.1.0/include #installed here by brew
		)

SET(CF4OCL2_LIB_SEARCH_PATHS
		/lib/
		/lib64/
		/usr/lib
		/usr/lib64
		/usr/local/lib
		/usr/local/lib64
		/usr/local/Cellar/cf4ocl/2.1.0/lib #installed here by brew
		)

FIND_PATH(CF4OCL2_INCLUDE_DIR NAMES cf4ocl2.h PATHS ${CF4OCL2_INCLUDE_SEARCH_PATHS})
FIND_LIBRARY(CF4OCL2_LIB NAMES cf4ocl2 PATHS ${CF4OCL2_LIB_SEARCH_PATHS})
message('CF4OCL2_LIB DIR:')
message(${CF4OCL2_LIB})
```

And Glib:

```
find_package(PkgConfig REQUIRED)
pkg_search_module(GLIB REQUIRED glib-2.0)
target_include_directories(<my target> PRIVATE ${GLIB_INCLUDE_DIRS})
target_link_libraries(<my target> INTERFACE ${GLIB_LDFLAGS})
```

Thank you for your time and just-in-time responses last day :)

**Comment by fakenmc** [comment]
Hi, thank you very much for the feedback. I'll leave this issue open so I can find a more general way to avoid this problem for future installations of cf4ocl by Mac users, if necessary by using your solution.
Best regards

**Comment by praisethemoon** [comment]
No problem, glad I could help :smiley:
